presto-main/src/main/java/com/facebook/presto/sql/planner/NullabilityAnalyzer.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner;
+
+import com.facebook.presto.sql.tree.Cast;
+import com.facebook.presto.sql.tree.DefaultExpressionTraversalVisitor;
+import com.facebook.presto.sql.tree.DereferenceExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.IfExpression;
+import com.facebook.presto.sql.tree.InPredicate;
+import com.facebook.presto.sql.tree.NullIfExpression;
+import com.facebook.presto.sql.tree.SearchedCaseExpression;
+import com.facebook.presto.sql.tree.SimpleCaseExpression;
+import com.facebook.presto.sql.tree.SubscriptExpression;
+import com.facebook.presto.sql.tree.TryExpression;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static java.util.Objects.requireNonNull;
+
+public final class NullabilityAnalyzer
+{
+    private NullabilityAnalyzer() {}
+
+    /**
+     * TODO: this currently produces a very conservative estimate.
+     * We need to narrow down the conditions under which certain constructs
+     * can return null (e.g., if(a, b, c) might return null for non-null a
+     * only if b or c can be null.
+     */
+    public static boolean mayReturnNullOnNonNullInput(Expression expression)
+    {
+        requireNonNull(expression, "expression is null");
+
+        AtomicBoolean result = new AtomicBoolean(false);
+        new Visitor().process(expression, result);
+        return result.get();
+    }
+
+    private static class Visitor
+            extends DefaultExpressionTraversalVisitor<Void, AtomicBoolean>
+    {
+        @Override
+        protected Void visitCast(Cast node, AtomicBoolean result)
+        {
+            result.set(node.isSafe()); // try_cast
+            return null;
```

Comment:
Cast from JSON to other types can return null on non-null input when the json string is "null".